File: Case1/draw_output_samples.py
# -*- coding: utf-8 -*-


#为了保持与第二个案例的绘图一致性，用matlab版本的draw_output_samples.m计算好结果，再用python绘图



import numpy as np
import matplotlib.pyplot as plt

N=100
time_index=10
time_step=10
obs_num=6

y_obs_prediction=np.loadtxt('y_obs_ensem_{0}_time_{1}.txt'.format(N,time_index))
obs_mean=np.loadtxt('y_obs_mean_{0}_time_{1}.txt'.format(N,time_index))
obs_true=np.loadtxt('true_obs_{0}_time_{1}.txt'.format(N,time_index))


#提取3个观测点的，每个时间步的数据
obs_index_ensem=[1,3,5]
fig,ax=plt.subplots(1,3,figsize=(19,6),squeeze=True)
for obs_index in range(len(obs_index_ensem)):
    obs_ensemble=np.zeros((time_step,N))
    for i in xrange(N):
        for t_index in xrange(time_step):
            obs_ensemble[t_index,i]=y_obs_prediction[obs_index_ensem[obs_index]+t_index*obs_num-1,i]
            
    obs_mean_selected=np.zeros(time_step)      
    for t_index in xrange(time_step):
        obs_mean_selected[t_index]=obs_mean[obs_index_ensem[obs_index]+t_index*obs_num-1]
        
    obs_true_selected=np.zeros(time_step)      
    for t_index in xrange(time_step):
        obs_true_selected[t_index]=obs_true[obs_index_ensem[obs_index]+t_index*obs_num-1]
    
    for k in xrange(N):
        if k==N-1:
            ax[obs_index].plot(range(time_step),obs_ensemble[:,k],c='b',alpha=0.2,label='Realizations')
            ax[obs_index].plot(range(time_step),obs_mean_selected,c='r',lw=1.5,alpha=1,label='Mean estimation')
            ax[obs_index].scatter(range(time_step),obs_true_selected,c='k',s=22,label='Observations',zorder=10)
        ax[obs_index].plot(range(time_step),obs_ensemble[:,k],c='b',alpha=0.1)
        min_y=np.int32(min(ax[obs_index].get_ylim()))
        max_y=np.int32(max(ax[obs_index].get_ylim()))
        ax[obs_index].set_ylim(min_y,max_y)
        ax[obs_index].set_yticks(np.linspace(min_y,max_y,6))
        ax[obs_index].set_yticklabels(np.int32(np.linspace(min_y,max_y,6)))
        
        
    ax[obs_index].plot(range(time_step),obs_mean_selected,c='r',lw=1.5,alpha=1)
    ax[obs_index].scatter(range(time_step),obs_true_selected,c='k',s=22)
    ax[obs_index].set_xlabel('Time step',fontsize=15)
    ax[obs_index].set_ylabel('Pressure(Pa)',fontsize=15)
    ax[obs_index].set_xlim(-1,11)
    plt.setp(ax[obs_index].get_xticklabels(),fontsize=14)
    plt.setp(ax[obs_index].get_yticklabels(),fontsize=14)




    
# ax[2].set_ylim(100800,101400)
# ax[2].set_yticks(np.linspace(100800,101400,6))
# ax[2].set_yticklabels(np.int32(np.linspace(100800,101400,6)))
plt.legend(loc='best')
plt.tight_layout() 
fig.subplots_adjust(wspace=0.25)
fig.savefig('PCIKF_fitting_map_time_{0}.jpeg'.format(time_index))  
plt.show()

File: Case1/draw_output_samples.m

N=100;
time_index=10;
time_step=10;
obs_num=6;

% para_sample=importdata(['./updated_para_sample_EnRML_1_',num2str(N),'_',num2str(time_index),'.mat']);
% para_mean=importdata(['./updated_para_mean_EnRML_1_',num2str(N),'_',num2str(time_index),'.mat']);

[para_sample,para_mean]=generate_PCIKF_post_para(time_index,N);

%计算样本输出
y_obs_ensem=zeros(obs_num*time_step,N);
parfor i=1:N
    y_obs_ensem(:,i)=ogs_call(para_sample(:,i),i,time_step);
end
save(['y_obs_ensem_',num2str(N),'_time_',num2str(time_index),'.txt'],'-ascii','y_obs_ensem');

% load(['y_obs_ensem_',num2str(N),'_time_',num2str(time_index),'.txt'])

%计算均值输出
y_obs_mean=ogs_call(para_mean,1,time_step);
save(['y_obs_mean_',num2str(N),'_time_',num2str(time_index),'.txt'],'-ascii','y_obs_mean')

%得到真实观测值
load para_true.mat
true_obs=ogs_call(para_true,999,time_step);
save(['true_obs_',num2str(N),'_time_',num2str(time_index),'.txt'],'-ascii','true_obs')



%提取3个观测点的每个时间步的观测数据,为了与案例2保持图例一致，转用draw_output_samples.py作图
obs_index_ensem=[1,3,5];

for obs_index=1:length(obs_index_ensem)
    obs_ensemble=zeros(time_step,N);
    for i=1:N
        for t_index=1:time_step
            obs_ensemble(t_index,i)=y_obs_ensem(obs_index_ensem(obs_index)+(t_index-1)*obs_num,i);
        end
    end
    
    obs_mean_selected=zeros(time_step,1);
    for t_index=1:time_step
        obs_mean_selected(t_index)=y_obs_mean(obs_index_ensem(obs_index)+(t_index-1)*obs_num);
    end
    
    obs_true_selected=zeros(time_step,1);
    for t_index=1:time_step
        obs_true_selected(t_index)=true_obs(obs_index_ensem(obs_index)+(t_index-1)*obs_num);
    end
    
    subplot(1,3,obs_index)
    for k=1:N
        plot(1:time_step,obs_ensemble(:,k),'b')
		hold on
    end
    plot(1:time_step,obs_mean_selected,'r-')
	hold on
    plot(1:time_step,obs_true_selected,'k*')
    
end